MBS-8952

sergioboi opened the pull request.

sergioboi committed:
MBS-8952 Update ScreenshotCapturer.kt
replace IllegalStateException with logging
before we threw exception when there was not a RESUMED Activity
And it makes ERROR log bigger and more unclear

sergioboi committed:
MBS-8952 Add FakeLogger.kt
replace all test's overrides of Logger.kt

sergioboi commented on subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/screenshot/ScreenshotCapturer.kt:
```diff
@@ -4,50 +4,81 @@ import android.app.Activity
+                    Option.Some(drawCanvas(activity))
+                }
+            } else {
+                logger.warn("There are no RESUMED activity when capturingBitmap")
```
### Before
<img width="1477" alt="Screenshot 2020-07-15 at 19 09 05" src="https://user-images.githubusercontent.com/8560287/87670922-34db9f00-c779-11ea-977e-0221b14cf6c6.png">

eugene-krivobokov replied:
```suggestion
                logger.warn("There is no RESUMED activity when capturingBitmap")
```
![](https://contenthub-static.grammarly.com/blog/wp-content/uploads/2016/11/THERE-IS-THERE-ARE.jpg)

sergioboi committed:
MBS-8952 Fix VideoCapturer.kt
add atomic state changing

sergioboi committed:
MBS-8952 Update VideoCapturer.kt
before exceptions in start and stop was unhandled,
it makes test finishing failed
fix concurrent start() and stop()

sergioboi commented on subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/video/VideoCaptureTestListener.kt:
```diff
@@ -49,8 +52,13 @@ class VideoCaptureTestListener(
             logger.debug("Video recording feature enabled. Recording starting")
-            videoCapturer.start()
-            logger.debug("Video recording feature enabled. Recording started")
+            when (val startingResult = videoCapturer.start()) {
```
Before: we could fail in `start()` and it led to unpredictable behavior

sergioboi commented on subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/video/VideoCaptureTestListener.kt:
```diff
@@ -63,23 +71,24 @@ class VideoCaptureTestListener(
-                logger.debug(
-                    "Video uploading enabled. Filed to upload video for ${state.testMetadata.className}.${state.testMetadata.methodName}"
-                )
+            when (val result = videoCapturer.stop()) {
```
Before we could fail in `stop` and it led to failure in `com.avito.android.test.report.ReportImplementation#reportTestCase` because of that we didn't call `com.avito.android.test.report.ReportImplementation#writeTestCase` where we write test result

sergioboi commented on subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/video/VideoCapturer.kt:
```diff
@@ -11,77 +11,104 @@ import com.avito.android.util.executeMethod
+}
 
-        executeRecorderCommand("stop", outputFile)
+class VideoCapturerImpl(
```
I handle all errors and try to reduce concurrent usage of `start` and `stop`
I was choosing between offensive and defensive implementations
And I consider video capturer to be not a blocker because we don't know how frequently recording start or stop may fail. Even one fail because of video recording may lead to PR rerun

sergioboi committed:
MBS-8952 Fix flakiness

eugene-krivobokov commented on subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/video/VideoCapturer.kt:
```diff
@@ -11,77 +11,105 @@ import com.avito.android.util.executeMethod
+                if (videoFile.exists()) {
+                    videoFile.delete()
+                }
+                VideoCapturer.StartingRecordResult.Error("Can't start capture", t)
```
```suggestion
                VideoCapturer.StartingRecordResult.Error("Can't start video capturing", t)
```

eugene-krivobokov commented on subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/video/VideoCapturer.kt:
```diff
@@ -11,77 +11,105 @@ import com.avito.android.util.executeMethod
+                VideoCapturer.StartingRecordResult.Error("Can't start capture", t)
+            }
+        } else {
+            VideoCapturer.StartingRecordResult.Error("Can't start capture capturer isn't Idling")
```
```suggestion
            VideoCapturer.StartingRecordResult.Error("Can't start video capturing: isn't Idling")
```

eugene-krivobokov commented on subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/video/VideoCapturer.kt:
```diff
@@ -11,77 +11,105 @@ import com.avito.android.util.executeMethod
+                } finally {
+                    if (videoFile.exists()) {
+                        videoFile.delete()
+                    }
```
Should we clean `outputFile` here like in `stop()`?

sergioboi replied:
We rewrite output file every time when invoke `start` and I decided to don't clear it

eugene-krivobokov commented on subprojects/android-test/ui-testing-core/src/main/kotlin/com/avito/android/test/util/Instrumentations.kt:
```diff
@@ -7,6 +7,10 @@ import androidx.test.runner.lifecycle.ActivityLifecycleMonitorRegistry
+    return getCurrentActivitySafe() ?: throw IllegalStateException("Resumed activity not found")
+}
+
+fun getCurrentActivitySafe(): Activity? {
```
```suggestion
fun getCurrentActivitySafely(): Activity? {
```
OR
```suggestion
fun getCurrentActivityOrNull(): Activity? {
```
`OrNull` is a commonly used naming pattern in such cases.

sergioboi replied:
Changed to OrNull

sergioboi committed:
MBS-8952 Fix ReportTestListener.kt
separate logging about SUCCESS and FAIL when processing events
move errors logging from Sentry to logcat
change level from Error to Warning because that isn't critical

sergioboi committed:
MBS-8952 Fix ReportImplementation.kt
Surround `afterTestStop` call with try catch
because before it led to unwritten test result

sergioboi commented on subprojects/android-test/test-inhouse-runner/src/main/kotlin/com/avito/android/runner/InHouseInstrumentationTestRunner.kt:
```diff
@@ -264,7 +267,7 @@ abstract class InHouseInstrumentationTestRunner :
             report.reportTestCase()
         } catch (t: Throwable) {
-            Log.e(tag, "Error during reporting test after global exception handling", t)
+            Log.w(tag, "Can't register and report unexpected incident", t)
```
We already don't fail a test if had a failure while register unexpected incident
But ERROR logging level led to misunderstanding behavior. So I change it to WARN

sergioboi commented on subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/ReportImplementation.kt:
```diff
@@ -196,49 +195,55 @@ class ReportImplementation(
-                currentState.endTime = timeProvider.nowInSeconds()
-                afterTestStop(this)
+            try {
+                afterTestStop(startedState)
```
Because of error in a test `Stop` callback test became unwritten
I prefer to log that Throwable to FAIL whole test

sergioboi commented on subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/ReportTestListener.kt:
```diff
@@ -12,59 +12,37 @@ class ReportTestListener : RunListener() {
-            )
-            report.sentry.sendException(t)
         }
-        Log.d(TAG, "TestFinished event processing completed for ${description.displayName}")
```
Before we log the event completed even if we had an error while processing. It made harder to realize what actually happen and how

sergioboi committed:
MBS-8952 Fix typo

sergioboi committed:
MBS-8952 Fix kotlinx html
we've lost kotlinx-html 0.6.9

sergioboi committed:
MBS-8952 Add JCenter Proxy for lost kotlinx-html

eugene-krivobokov commented on settings.gradle.kts:
```diff
@@ -67,6 +77,20 @@ pluginManagement {
+                    }
+                }
+                filter {
+                    includeVersion("org.jetbrains.kotlinx", "kotlinx-html-jvm", "0.6.9")
```
:+1:
